### DISCUSSION TITLE
How to validate a route without sending a request to an API?

### DISCUSSION THREAD

**Discussion by ghost** [nuxt/nuxt#18869] [Questions] [answered]
Fyi Nuxt is in SSG mode (`generate`).
I am using the following code in my `/pages/projects/[project].vue` page:

```vue
<script setup>
definePageMeta({
    async validate(route) {
        const slug = route.params.project;
        const projectSlugs = await useProjectSlugs();
        const matchingRoute = item => item.slug === slug;
        return projectSlugs.value.some(matchingRoute);
    }
});
</script>
```

The `useProjectSlugs()` composable fetches a list of all available/allowed/public slugs from the API.
It works great, but every time I enter the route an API request is sent. I don’t think that this is really necessary, especially when using in SSG mode.

Ideally it would make sense to get the `projectSlugs` from a prerendered value/store (like data in `_payload.js`) to prevent an API call. The data I want to use for validating is already there (in payload.js files), …

Is there a way to access the data from payload.js to prevent these extra API calls?

**Comment by TheAlexLichter** [comment]
Hmm. You could e.g. save them in a pinia store in you `app.vue` and treat it as a source then

**Reply by ghost** [reply]
Why pinia and not `useStore`? And what do you mean by "treat it as a source then"?

**Comment by ghost** [comment, marked as answer]
I’ve figured out a great solution myself. And I wanted to share it with you.

I’ve written a plugin `/plugins/slugs.js` with this code:

```js
export default defineNuxtPlugin(async (nuxtApp) => {
    const projectSlugs = process.client
        ? ref(nuxtApp.payload.data["project-slugs"] ?? [])
        : await useProjectSlugs();

    return { provide: { projectSlugs } };
});
```

This plugin makes sure that the data fetched on the server side is stored inside the global _payload.js (with `await useProjectSlugs()`). It also makes sure to use the cached payload data client side (with `nuxtApp.payload.data["project-slugs"]`), instead of calling the API again. Use this plugin inside a validate method (middleware), to validate the route by the provided slugs.

So in my `/pages/projects/[project].vue` I am using the following to validate the route:

```vue
<script setup>
definePageMeta({
    async validate(route) {
        const { $projectSlugs } = useNuxtApp();
        const slug = route.params.project;
        const matchingRoute = item => item.slug === slug;
        return $projectSlugs.value.some(matchingRoute);
    }
});
</script>
```

`/composables/useProjectSlugs.js` looks something like this:

```js
export const useProjectSlugs = async () => {
    const { data } = await useAsyncData("project-slugs", async () => {
        const items = await YOUR_API_CALL();
        return items;
    });
    return data;
};

```

**Reply by ghost** [reply]
@manniL What do you think?
